quantecon/game_theory/polymatrix_game.py

```diff
@@ -0,0 +1,317 @@
+    that try to sum to the payoff.
+    Precise when the game can be represented with a polymatrix;
+    otherwise, an approximation using least squares on the
+    payoffs at every action combination.
+    If an approximation, it may drastically change the
+    game, use with caution.
+
+    Parameters
+    ----------
+    nfg : NormalFormGame
+        The game.
+
+    my_player_number : int
+        The number of the player making the action.
+
+    my_action_number : int
+        The number of our player's action.
+
+    is_polymatrix : bool, optional
+        Is the game represented by this normal form
+        actually a polymatrix game. Defaults to True.
+
+    Returns
+    -------
+    dict[tuple[int, int], float]
+        Dictionary giving a (approximate) component
+        of the payoff at each other player and their action.
+    """
+    action_combinations = product(*(
+        [
+            range(nfg.nums_actions[p]) if p != my_player_number
+            else [my_action_number]
+            for p in range(nfg.N)
+        ]
+    ))
+    my_player: Player = nfg.players[my_player_number]
+    my_payoffs = my_player.payoff_array
+    hh_actions_and_payoffs = np.vstack([
+        np.hstack(
+            [
+                np.eye(nfg.nums_actions[p])[action_combination[p]]
+                for p in range(nfg.N) if p != my_player_number
+            ] + [
+                my_payoffs[
+                    action_combination[my_player_number:]
+                    + action_combination[:my_player_number]
+                ]
+            ]
+        )
+        for action_combination in action_combinations
+    ])
+    hh_actions = hh_actions_and_payoffs[:, :-1]
+    combined_payoffs = hh_actions_and_payoffs[:, -1]
+
+    # Could use pinv, but this is clearer.
+    hh_payoffs_array, _, _, _ = np.linalg.lstsq(
+        hh_actions,
+        combined_payoffs,
+        rcond=None
+    )
+    if is_polymatrix:
+        was_polymatrix = np.allclose(
+            hh_actions @ hh_payoffs_array,
+            combined_payoffs
+        )
+        assert was_polymatrix, "The game is not actually a polymatrix game."
+
+    payoff_labels = [
+        (p, a)
+        for p in range(nfg.N) if p != my_player_number
+        for a in range(nfg.nums_actions[p])
+    ]
+
+    payoffs = {
+        label: payoff
+        for label, payoff in zip(payoff_labels, hh_payoffs_array)}
+
+    return payoffs
+
+
+class PolymatrixGame:
+    """
+    Polymatrix Game.
+
+    `polymatrix[(a, b)]` is a 2D matrix, the payoff matrix of `a`
+    in the bimatrix game between `a` and `b`;
+    player number `a` is the row player
+    and player number `b` is the column player.
+
+    Attributes
+    ----------
+    N : int
+        Number of players.
+
+    nums_actions : Sequence[int]
+        The number of actions available to each player.
+
+    polymatrix : Mapping[ tuple[int, int], Matrix ]
+        Maps each pair of player numbers to a matrix.
```

Comment:
```suggestion
    Attributes
    ----------
    N : scalar(int)
        Number of players.

    nums_actions : tuple(int)
        The number of actions available to each player.

    polymatrix : dict[tuple(int), ndarray(float, ndim=2)]
        Maps each pair of player numbers to a matrix.
```

(And other places)

Reply:
Refreshed and saw this after committing.
The convention I have gone for throughout (except for the `ndim=2` part) is like `tuple[int]` instead of `tuple(int)` because this is what VS code generates automatically for the numpy style docs. However, I think it is alright if sometimes integers are documented as `int` and other times as `scalar(int)`.

Reply:
I made the Sequence instead of concrete type on returns and attributes a few times though. Thanks.